### ISSUE TITLE
take + repeat + distinct not completing

### ISSUE THREAD

**Issue by hamen** [ReactiveX/RxJava#2587]
Hi all
I'm having a hard time trying to figure this out:

```java
Observable<AppInfo> sequence = Observable.from(apps)
                .take(3)
                .repeat(3);

        sequence.subscribe(new Observer<AppInfo>() {
            @Override
            public void onCompleted() {
                Log.d("RXJAVA1", "Completed");
            }

            @Override
            public void onError(Throwable e) {
                Log.e("RXJAVA1", "Error");
            }

            @Override
            public void onNext(AppInfo appInfo) {
                Log.d("RXJAVA1", appInfo.getName());
            }
        });

        sequence.distinct()
                .subscribe(new Observer<AppInfo>() {
                    @Override
                    public void onCompleted() {
                        Log.d("RXJAVA2", "Completed");
                    }

                    @Override
                    public void onError(Throwable e) {
                        Log.e("RXJAVA2", "Error");
                    }

                    @Override
                    public void onNext(AppInfo appInfo) {
                        Log.d("RXJAVA2", appInfo.getName());
                    }
                });
```
gives me
```
D/RXJAVA1﹕ 9GAG
D/RXJAVA1﹕ AdSense
D/RXJAVA1﹕ AirDroid
D/RXJAVA1﹕ 9GAG
D/RXJAVA1﹕ AdSense
D/RXJAVA1﹕ AirDroid
D/RXJAVA1﹕ 9GAG
D/RXJAVA1﹕ AdSense
D/RXJAVA1﹕ AirDroid
D/RXJAVA1﹕ Completed
D/RXJAVA2﹕ 9GAG
D/RXJAVA2﹕ AdSense
D/RXJAVA2﹕ AirDroid
```
and the sequence never completes. If I only repeat it twice:

```java
Observable<AppInfo> sequence = Observable.from(apps)
                .take(3)
                .repeat(2);

        sequence.subscribe(new Observer<AppInfo>() {
            @Override
            public void onCompleted() {
                Log.d("RXJAVA1", "Completed");
            }

            @Override
            public void onError(Throwable e) {
                Log.e("RXJAVA1", "Error");
            }

            @Override
            public void onNext(AppInfo appInfo) {
                Log.d("RXJAVA1", appInfo.getName());
            }
        });

        sequence.distinct()
                .subscribe(new Observer<AppInfo>() {
                    @Override
                    public void onCompleted() {
                        Log.d("RXJAVA2", "Completed");
                    }

                    @Override
                    public void onError(Throwable e) {
                        Log.e("RXJAVA2", "Error");
                    }

                    @Override
                    public void onNext(AppInfo appInfo) {
                        Log.d("RXJAVA2", appInfo.getName());
                    }
                });
```
I get
```
D/RXJAVA1﹕ 9GAG
D/RXJAVA1﹕ AdSense
D/RXJAVA1﹕ AirDroid
D/RXJAVA1﹕ 9GAG
D/RXJAVA1﹕ AdSense
D/RXJAVA1﹕ AirDroid
D/RXJAVA1﹕ Completed
D/RXJAVA2﹕ 9GAG
D/RXJAVA2﹕ AdSense
D/RXJAVA2﹕ AirDroid
D/RXJAVA2﹕ Completed
```
and everything is fine.
I'm working on some examples and I spotted this weird scenario and it's driving me crazy. Am I missing something or there is actually something wrong with this?
Thanks

**Comment by akarnokd** [comment]
Hello. It is quite odd. The following program works for me on master:
```java
public class Issue2587 {
    static void logD(String key, Object msg) {
        System.out.print(key);
        System.out.print(" ");
        System.out.println(msg);
    }
    public static void main(String[] args) {
        Observable<Integer> sequence = Observable.from(Arrays.asList(1, 2, 3, 4, 5))
                .take(3)
                .repeat(3);

        sequence.subscribe(new Observer<Integer>() {
            @Override
            public void onCompleted() {
                logD("RXJAVA1", "Completed");
            }

            @Override
            public void onError(Throwable e) {
                logD("RXJAVA1", "Error");
            }

            @Override
            public void onNext(Integer appInfo) {
                logD("RXJAVA1", appInfo);
            }
        });
    }
}
```
Prints:
```
RXJAVA1 1
RXJAVA1 2
RXJAVA1 3
RXJAVA1 1
RXJAVA1 2
RXJAVA1 3
RXJAVA1 1
RXJAVA1 2
RXJAVA1 3
RXJAVA1 Completed
```

**Comment by hamen** [comment]
Could you try to apply `distinct()` to get rid of the duplicates please?

**Comment by akarnokd** [comment]
Hmm. If I apply distinct(), I don't get the Completed message either. I'm going to investigate this.

**Comment by akarnokd** [comment]
It is a bug with repeat. I'll post the fix soon.

**Comment by akarnokd** [comment]
Could you verify #2589 works for you?

**Comment by akarnokd** [comment]
The fix is available in 1.0.5. Thanks for reporting.

**Comment by hamen** [comment]
Sorry for the delay due to the timezone.
I checked the `1.0.5`: everything is working fine.
Thank you for the awesome response!
